presto-main/src/main/java/io/prestosql/metadata/MetadataManager.java

```diff
@@ -390,30 +398,39 @@ public boolean catalogExists(Session session, String catalogName)
             return Optional.empty();
         }
 
-        return Optional.of(new TableLayoutResult(fromConnectorLayout(connectorId, transaction, layouts.get(0).getTableLayout()), layouts.get(0).getUnenforcedConstraint()));
+        ConnectorTableLayout tableLayout = layouts.get(0).getTableLayout();
+        return Optional.of(new TableLayoutResult(
+                new TableHandle(connectorId, connectorTable, transaction, Optional.of(tableLayout.getHandle())),
+                new TableLayout(connectorId, transaction, tableLayout),
+                layouts.get(0).getUnenforcedConstraint()));
     }
 
     @Override
-    public TableLayout getLayout(Session session, TableLayoutHandle handle)
+    public TableLayout getLayout(Session session, TableHandle handle)
     {
         ConnectorId connectorId = handle.getConnectorId();
         CatalogMetadata catalogMetadata = getCatalogMetadata(session, connectorId);
         ConnectorMetadata metadata = catalogMetadata.getMetadataFor(connectorId);
-        ConnectorTransactionHandle transaction = catalogMetadata.getTransactionHandleFor(connectorId);
-        return fromConnectorLayout(connectorId, transaction, metadata.getTableLayout(session.toConnectorSession(connectorId), handle.getConnectorHandle()));
+        ConnectorSession connectorSession = session.toConnectorSession(connectorId);
+
+        return handle.getLayout()
+                .map(layout -> new TableLayout(connectorId, handle.getTransaction(), metadata.getTableLayout(connectorSession, layout)))
+                .orElseGet(() -> getLayout(session, handle, Constraint.alwaysTrue(), Optional.empty())
```

Comment:
It makes me wonder if `TableHandle` shouldn't have basic layout set at planning time so that layout cannot be absent. Then layout could only be improved later on by pushing down predicates.

Reply:
I went down that path at some point but it was a dead end. It certainly would have made things easier. The problem is that choosing a layout naively (without filter) causes the hive connector to try to enumerate all partitions in the table.

Reply:
Could we insert some "dummy" (noop) table handle, but I guess it won't be better than just `Optional.empty()`.

Reply:
Nah, that requires a bunch of special-casing everywhere and doesn't simplify things much.